lib/ddtrace/contrib/active_support/notifications.rb

```diff
@@ -0,0 +1,40 @@
+module Datadog
+  module Contrib
+    module ActiveSupport
+      class Notifications
+        def self.subscribe(pattern, span_name, options = {}, tracer = Datadog.tracer, &block)
+          subscriber = Subscriber.new(tracer, span_name, options, &block)
+
+          ::ActiveSupport::Notifications.subscribe(pattern, subscriber)
+        end
+
+        class Subscriber
+          def initialize(tracer, span_name, options, &block)
+            @tracer = tracer
+            @span_name = span_name
+            @options = options
+            @block = block
+
+            # A stack of open spans. We want to get to the most recently opened span
+            # at any point in time.
+            @spans = []
+          end
+
+          def start(_name, _id, _payload)
+            @spans << @tracer.trace(@span_name, @options)
+          end
+
+          def finish(name, id, payload)
+            # We close spans in reverse order.
+            span = @spans.pop or raise "no spans left in stack!"
```

Comment:
I'm a bit concerned using a stack here would be thread unsafe. You mentioned getting the `@tracer.active_span` didn't work; what was the issue you ran into?

Reply:
I'm not sure – the spans aren't completed correctly. I'd love to get some feedback on that.

Reply:
Yeah, I think I'd like to take a look at that. I'll see if I can clone and test this.

Reply:
@dasch Okay, I cloned down your changes locally, and dug deeper into this.

This `@tracer.active_span` wasn't working as the only change because during the tests, the `unpatch` method was creating many subscriptions with many tracers. If I was understanding the issue correctly, when there are multiple tracers concurrently writing, it tends to cause problems with context propagation, which was causing the tests to fail.

By applying the `@tracer.active_span` change in conjunction with the `subscribers` change suggested in the other comment thread, I got these tests passing locally. I've created a patch file which has all the above changes, which will fix the tests and Rubocop issues breaking the CI build. Feel free to try it out on your branch!

Git diff (apply on your branch with `git apply datadog_asn.txt`):
[datadog_asn.txt](https://github.com/DataDog/dd-trace-rb/files/1709122/datadog_asn.txt)